File: fullstack-example-frontend/todoapp/src/app/common/models/todo-item.ts
import { MatrixY } from './matrix-y';
import { MatrixX } from './matrix-x';
import { isDateEqual } from '../date-utility';
import { DateTime } from 'luxon';

export class TodoItem {
    public id: string;
    public listId: string | null;
    public name: string;
    public matrixX: MatrixX;
    public matrixY: MatrixY;
    public note: string;
    public dueDate: DateTime | null;
    public createdAt: DateTime;
    public modifiedAt: DateTime;
    public deletedAt: DateTime | null;
    public markedAsDone: boolean;

    constructor(
        id: string,
        listId: string | null,
        name: string,
        matrixX: MatrixX,
        matrixY: MatrixY,
        note: string,
        dueDate: DateTime | null,
        createdAt: DateTime,
        modifiedAt: DateTime,
        deletedAt: DateTime | null,
        markedAsDone: boolean
    ) {
        this.id = id;
        this.listId = listId;
        this.name = name;
        this.matrixX = matrixX;
        this.matrixY = matrixY;
        this.note = note;
        this.dueDate = dueDate;
        this.createdAt = createdAt;
        this.modifiedAt = modifiedAt;
        this.deletedAt = deletedAt;
        this.markedAsDone = markedAsDone;
    }

    public equals(other: TodoItem): boolean {
        return this.id === other.id &&
        this.listId === other.listId &&
        this.name === other.name &&
        this.matrixX === other.matrixX &&
        this.matrixY === other.matrixY &&
        this.note === other.note &&
        isDateEqual(this.dueDate, other.dueDate) &&
        isDateEqual(this.createdAt, other.createdAt) &&
        isDateEqual(this.modifiedAt, other.modifiedAt) &&
        isDateEqual(this.deletedAt, other.deletedAt) &&
        this.markedAsDone === other.markedAsDone;
    }
}

File: fullstack-example-backend/src/Eisenhower.Todo.Domain/Models/TodoItem.cs
namespace Eisenhower.Todo.Domain;

// TODO: implement IEquatable
public class TodoItem
{
    public TodoItemId TodoItemId { get; private set; }
    public UserId UserId { get; private set; }
    public TodoListId TodoListId { get; private set; }
    public string Name { get; private set;}
    public MatrixX MatrixX { get; private set; }
    public MatrixY MatrixY { get; private set; }
    public string Note { get; private set; }
    public bool MarkedAsDone { get; private set; }
    public DateTime? DueDate { get; private set; }
    public DateTime CreatedAt { get; private set; }
    public DateTime ModifiedAt { get; private set; }
    public DateTime? DeletedAt { get; private set; }

    public TodoItem(
        TodoItemId id,
        UserId userId,
        TodoListId listId,
        string name,
        MatrixX matrixX,
        MatrixY matrixY,
        string note,
        bool markedAsDone,
        DateTime? duedate,
        DateTime createdAt,
        DateTime modifiedAt,
        DateTime? deletedAt
    ) {
        if (string.IsNullOrEmpty(name))
            throw new ArgumentNullException(nameof(name));

        this.TodoItemId =  id;
        this.UserId = userId;
        this.TodoListId = listId;
        this.Name = name;
        this.MatrixX = matrixX;
        this.MatrixY = matrixY;
        this.Note = note;
        this.MarkedAsDone = markedAsDone;
        this.DueDate = duedate;
        this.CreatedAt = createdAt;
        this.ModifiedAt = modifiedAt;
        this.DeletedAt = deletedAt;
    }
}
